auth/auth.go

```
@@ -249,31 +262,151 @@ func (auth *Authenticator) Save(p Principal) error {
+func (auth *Authenticator) UpdateUserEmail(u User, email string) error {
+
+	updateUserEmailCallback := func(currentPrincipal Principal) (updatedPrincipal Principal, err error) {
+		currentUser, ok := currentPrincipal.(User)
+		if !ok {
+			return nil, base.ErrUpdateCancel
+		}
+
+		if currentUser.Email() == email {
+			return currentUser, base.ErrUpdateCancel
+		}
+
+		base.Debugf(base.KeyAuth, "Updating user %s email to: %v", base.UD(u.Name()), base.UD(email))
+		err = currentUser.SetEmail(email)
+		if err != nil {
+			return nil, err
 		}
+		return currentUser, nil
 	}
+
+	updateError := auth.casUpdatePrincipal(u, updateUserEmailCallback)
+	if updateError != nil {
+		base.Infof(base.KeyAuth, "Unable to update user email.  User:%s Error:%v", base.UD(u.Name()), updateError)
+	}
+	return updateError
+}
+
+// rehashPassword will check the bcrypt cost of the given hash
+// and will reset the user's password if the configured cost has since changed
+// Callers must verify password is correct before calling this
+func (auth *Authenticator) rehashPassword(user User, password string) error {
+
+	// Exit early if bcryptCost has not been set
+	if !bcryptCostChanged {
+		return nil
+	}
+	var hashCost int
+	rehashPasswordCallback := func(currentPrincipal Principal) (updatedPrincipal Principal, err error) {
+
+		currentUserImpl, ok := currentPrincipal.(*userImpl)
+		if !ok {
+			return nil, base.ErrUpdateCancel
+		}
+
+		hashCost, costErr := bcrypt.Cost(currentUserImpl.PasswordHash_)
+		if costErr == nil && hashCost != bcryptCost {
+			// the cost of the existing hash is different than the configured bcrypt cost.
+			// We'll re-hash the password to adopt the new cost:
+			currentUserImpl.SetPassword(password)
+			return currentUserImpl, nil
+		} else {
+			return nil, base.ErrUpdateCancel
+		}
+	}
+
+	err := auth.casUpdatePrincipal(user, rehashPasswordCallback)
+	if err != nil {
+		base.Warnf(base.KeyAll, "Unable to save user when rehashing password: %v", err)
+		return err
```

Would there be any advantage in absorbing this error and just logging a warning?  Eg, treat the rehash op as a "best effort" attempt, and if it fails, don't change any other control flow by returning an error?

---

I think we need to return the error, and let the caller decide whether it's critical or not.  In this case the caller is already ignoring the error.  I think it makes sense to move this logging up the stack, though.